ethcore/src/engines/authority_round/mod.rs

```diff
@@ -125,54 +148,87 @@ impl From<ethjson::spec::AuthorityRoundParams> for AuthorityRoundParams {
+	/// overflow.
 	fn duration_remaining(&self) -> Duration {
-		let now = unix_now();
-		let expected_seconds = self.load()
-			.checked_add(1)
-			.and_then(|ctr| ctr.checked_mul(self.duration as u64))
-			.map(Duration::from_secs);
-
-		match expected_seconds {
-			Some(step_end) if step_end > now => step_end - now,
-			Some(_) => Duration::from_secs(0),
-			None => {
-				let ctr = self.load();
-				error!(target: "engine", "Step counter is too high: {}, aborting", ctr);
-				panic!("step counter is too high: {}", ctr)
-			},
-		}
+		self.opt_duration_remaining().unwrap_or_else(|| {
+			let ctr = self.load();
+			error!(target: "engine", "Step counter under- or overflow: {}, aborting", ctr);
+			panic!("step counter under- or overflow: {}", ctr)
+		})
+	}
 
+	/// Finds the remaining duration of the current step. Returns `None` if there was a counter
+	/// under- or overflow.
+	fn opt_duration_remaining(&self) -> Option<Duration> {
+		let next_step = self.load().checked_add(1)?;
+		let (mut prev_dur, mut prev_step, mut prev_time) =
+			self.durations.first().expect("durations cannot be empty");
+		for (step, time, dur) in self.durations.iter().skip(1) {
+ 			if *step >= next_step {
+				break;
+			}
+			prev_step = *step;
+			prev_time = *time;
+			prev_dur = *dur;
+		}
+		let time = prev_time.checked_add(next_step.checked_sub(prev_step)?.checked_mul(prev_dur)?)?;
+		Some(Duration::from_secs(time.saturating_sub(unix_now().as_secs())))
 	}
 
+	/// Increments the step number.
+	///
+	/// Panics if the new step number is `u64::MAX`.
 	fn increment(&self) {
-		use std::usize;
 		// fetch_add won't panic on overflow but will rather wrap
 		// around, leading to zero as the step counter, which might
 		// lead to unexpected situations, so it's better to shut down.
-		if self.inner.fetch_add(1, AtomicOrdering::SeqCst) == usize::MAX {
-			error!(target: "engine", "Step counter is too high: {}, aborting", usize::MAX);
-			panic!("step counter is too high: {}", usize::MAX);
+		if self.inner.fetch_add(1, AtomicOrdering::SeqCst) == u64::MAX {
+			error!(target: "engine", "Step counter is too high: {}, aborting", u64::MAX);
+			panic!("step counter is too high: {}", u64::MAX);
 		}
-
 	}
 
 	fn calibrate(&self) {
 		if self.calibrate {
-			let new_step = unix_now().as_secs() / (self.duration as u64);
-			self.inner.store(new_step as usize, AtomicOrdering::SeqCst);
+			if self.opt_calibrate().is_none() {
+				let ctr = self.load();
+				error!(target: "engine", "Step counter under- or overflow: {}, aborting", ctr);
+				panic!("step counter under- or overflow: {}", ctr)
+			}
 		}
 	}
 
+	/// Calibrates the AuRa step number given according to the current time.
+	fn opt_calibrate(&self) -> Option<()> {
+		let now = unix_now().as_secs();
+		let (mut prev_step, mut prev_time, mut prev_dur) =
+			self.durations.first().expect("durations cannot be empty");
+		for (step, time, dur) in self.durations.iter().skip(1) {
+			if *time >= now {
```

Comment:
Should this actually be strict? If `time == now`, I imagine we consider the new step to have started?

Reply:
You are right. `==` works better for `Vec`.